instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+module OpenTelemetry
+  module Instrumentation
+    # This class contains the ASN subsciber that instruments message processing
+    class ProcessMessageSubscriber
+      def tracer
+        Racecar::Instrumentation.instance.tracer
+      end
+
+      def start(_name, _id, payload)
+        attrs = attributes(payload)
+
+        parent_context = OpenTelemetry.propagation.extract(payload[:headers], getter: OpenTelemetry::Common::Propagation.symbol_key_getter)
+        parent_token = OpenTelemetry::Context.attach(parent_context)
+
+        span_context = OpenTelemetry::Trace.current_span(parent_context).context
+        links = [OpenTelemetry::Trace::Link.new(span_context)] if span_context.valid?
+
+        span = tracer.start_span("#{payload[:topic]} process", kind: :consumer, attributes: attrs, links: links)
+        token = OpenTelemetry::Context.attach(OpenTelemetry::Trace.context_with_span(span))
+        payload.merge!(
+          __opentelemetry_span: span,
+          __opentelemetry_ctx_token: token,
+          __opentelemetry_parent_ctx_token: parent_token
+        )
+      end
+
+      def attributes(payload)
+        attributes = {
+          'messaging.system' => 'kafka',
+          'messaging.destination' => payload[:topic],
+          'messaging.destination_kind' => 'topic',
+          'messaging.kafka.partition' => payload[:partition],
+          'messaging.kafka.offset' => payload[:offset]
+        }
+
+        attributes['messaging.kafka.message_key'] = payload[:key] if payload[:key]
+        attributes
+      end
+
+      def finish(name, id, payload)
```

Comment:
I think this instrumentation is adapted from the `ActiveSupport` instrumentation's `SpanSubscriber` ([link](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/22796cb8141b3100c6ce94c25a15e31df3313464/instrumentation/active_support/lib/opentelemetry/instrumentation/active_support/span_subscriber.rb)), so maybe these questions were answered over there. At the risk of being told to RTFM, are we sure that:

1. `start` and `finish` will be called sequentially, with all context.attach calls matched with a context.detach call.
2. `start` and `finish` will be invoked by the same `Thread`.

If user code attaches to a new context between `start` and `finish`, but neglects to detach from that context, then [`detach`](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/api/lib/opentelemetry/context.rb#L55-L62) will not work as expected. Of course, they'd see an error in their logs, so ... that'll probably help them fix it.

Similarly, if `finish` is called in a different thread than `start`, it'll be using the wrong Context stack, since that's stored on [`Thread.current`](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/api/lib/opentelemetry/context.rb#L127).

I don't know much about ActiveSupport notifications (or Rdkafka! or Racecar!), so I'm most likely being paranoid here, but thought it was worth asking.

Reply:
@ahayworth might have insight, since he worked on a [similar thing](https://github.com/open-telemetry/opentelemetry-ruby/pull/796).

Reply:
TBH, it's not 100% in manual, but `ActiveSupport::Notifications` will call in that order and in the same thread (they leave to subscribers to decide if async is needed). The code isn't terribly complicated [so you could follow on the source](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/notifications.rb)